libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Runtime/Adapters/MultiplexingAdapter.cs

```diff
@@ -0,0 +1,170 @@
+        {
+            return GetAppropriateAdapter(channelId).CreateConversationAsync(botAppId, channelId, serviceUrl, audience, conversationParameters, callback, cancellationToken);
+        }
+
+        /// <inheritdoc/>
+        public override Task<InvokeResponse> ProcessActivityAsync(ClaimsIdentity claimsIdentity, Activity activity, BotCallbackHandler callback, CancellationToken cancellationToken)
+        {
+            return GetAppropriateAdapter(activity.ChannelId).ProcessActivityAsync(claimsIdentity, activity, callback, cancellationToken);
+        }
+
+        /// <summary>
+        /// Maps the incoming ChannelId or route (api/{route}) to the correct adapter. By default, this uses the AdapterSettings specified in appsettings.json.
+        /// Override this method to customize the mapping behavior.
+        /// </summary>
+        /// <param name="adapters">The adapters to map the routes to.</param>
+        protected virtual void BuildAdapterMap(IEnumerable<IBotFrameworkHttpAdapter> adapters)
+        {
+            var adapterSettings = _configuration.GetSection(AdapterSettings.AdapterSettingsKey).Get<List<AdapterSettings>>() ?? new List<AdapterSettings>();
+            var messagesAdapter = adapterSettings.FirstOrDefault(s => s.Route == "messages");
+            if (messagesAdapter == null)
+            {
+                adapterSettings.Add(AdapterSettings.CoreBotAdapterSettings);
+            }
+
+            foreach (var adapter in adapters ?? throw new ArgumentNullException(nameof(adapters)))
+            {
+                var settings = adapterSettings.FirstOrDefault(s => s.Enabled && s.Type == adapter.GetType().FullName.Split('.').Last());
+
+                if (settings != null)
+                {
+                    // Map route/ChannelId to adapter.
+                    _adapters.Add(settings.Route, adapter);
+
+                    // Add the adapter handling the messages route as the default adapter.
+                    if (settings.Route == "messages")
+                    {
+                        _adapters.Add("default", adapter);
+                    }
+                }
+            }
+        }
+
+        private void Initialize(IEnumerable<IBotFrameworkHttpAdapter> adapters)
+        {
+            BuildAdapterMap(adapters);
+        }
+
+        private BotAdapter GetAppropriateAdapter(string channelId)
+        {
+            if (_adapters.TryGetValue(channelId, out var adapter))
+            {
+                return adapter as BotAdapter;
+            }
+
+            if (_adapters.TryGetValue("default", out var defaultAdapter))
+            {
+                _logger.LogWarning($"Unable to find router for channelId {channelId}. Using default adapter {nameof(defaultAdapter)}. You may need to ensure your appsettings.json has an adapter with the Route property that matches the ChannelId of the adapter.");
+                return defaultAdapter as BotAdapter;
+            }
```

Comment:
Note: I'm okay removing this warning. `BuildAdapterMap()` maps an adapter to the `messages` route so this warning should occur pretty rarely, but I could see it maybe being annoying.